src/NumericalAlgorithms/Integration/GslQuadAdaptive.hpp

```diff
@@ -0,0 +1,290 @@
+ * many function evaluations take place in each subinterval. A higher-order rule
+ * serves better for smooth functions, whereas a lower-order rule saves time for
+ * functions with local difficulties, such as discontinuities.
+ */
+template <>
+class GslQuadAdaptive<GslIntegralType::StandardGaussKronrod>
+    : public detail::GslQuadAdaptiveImpl {
+ public:
+  using detail::GslQuadAdaptiveImpl::GslQuadAdaptiveImpl;
+  template <typename IntegrandType>
+  double operator()(IntegrandType&& integrand, const double lower_boundary,
+                    const double upper_boundary, const double tolerance_abs,
+                    const int key, const double tolerance_rel = 0.) const
+      noexcept {
+    double result;
+    double error;
+    detail::disable_gsl_error_handling();
+    const int status_code = gsl_integration_qag(
+        gsl_integrand(std::forward<IntegrandType>(integrand)), lower_boundary,
+        upper_boundary, tolerance_abs, tolerance_rel, this->max_intervals_, key,
+        this->workspace_.get(), &result, &error);
+    detail::check_status_code(status_code);
+    return result;
+  }
+};
+
+/*!
+ * \brief Integrates a 1D-function with singularities
+ *
+ * The algorithm for "IntegrableSingularitiesPresent" concentrates new,
+ * increasingly smaller subintervals around an unknown singularity and makes
+ * successive approximations to the integral which should converge towards a
+ * limit. The integration region is defined by `lower_boundary` and
+ * `upper_boundary`.
+ */
+template <>
+class GslQuadAdaptive<GslIntegralType::IntegrableSingularitiesPresent>
+    : public detail::GslQuadAdaptiveImpl {
+ public:
+  using detail::GslQuadAdaptiveImpl::GslQuadAdaptiveImpl;
+  template <typename IntegrandType>
+  double operator()(IntegrandType&& integrand, const double lower_boundary,
+                    const double upper_boundary, const double tolerance_abs,
+                    const double tolerance_rel = 0.) const noexcept {
+    double result;
+    double error;
+    detail::disable_gsl_error_handling();
+    const int status_code = gsl_integration_qags(
+        gsl_integrand(std::forward<IntegrandType>(integrand)), lower_boundary,
+        upper_boundary, tolerance_abs, tolerance_rel, this->max_intervals_,
+        this->workspace_.get(), &result, &error);
+    detail::check_status_code(status_code);
+    return result;
+  }
+};
+
+/*!
+ * \brief Integrates a 1D-function where singularities are known
+ *
+ * The algorithm for "IntegrableSingularitiesKnown" uses user-defined
+ * subintervals given by a vector of doubles `points`, where each element
+ * denotes an interval boundary.
+ */
+template <>
+class GslQuadAdaptive<GslIntegralType::IntegrableSingularitiesKnown>
+    : public detail::GslQuadAdaptiveImpl {
+ public:
+  using detail::GslQuadAdaptiveImpl::GslQuadAdaptiveImpl;
+  template <typename IntegrandType>
+  double operator()(IntegrandType&& integrand, const std::vector<double> points,
```

Comment:
@nilsdeppe should `points` be taken as const-reference?

Reply:
Yes, please make it `const std::vector<double>&`

Reply:
ping on this